### PR TITLE
Scrapper, Darya Balba - 21FPL2

### PR DESCRIPTION
**Pull request by daryabalba**

### CODE REVIEW

**Review by demid5111 — CHANGES_REQUESTED** [review]
Great that you have created PR. It is time to start working on it, To this moment, nothing to review.

**Comment by demid5111** [comment]
nothing to review. hurry up!

File: seminars/main.py
Comment (demid5111) [inline seminars/main.py:1]: discard unrelated changes

File: seminars/Text_response.txt
Comment (demid5111) [inline seminars/Text_response.txt:1]: discard unrelated changes

File: seminars/Text_response.html
Comment (demid5111) [inline seminars/Text_response.html:1]: discard unrelated changes

File: lab_5_scrapper/target_score.txt [hunk 1]
```diff
@@ -1 +1 @@
-6
\ No newline at end of file
+4
```
Comment (demid5111) [inline lab_5_scrapper/target_score.txt]: too small for you...

File: lab_5_scrapper/scrapper_config.json [hunk 2]
```diff
@@ -1,9 +1,9 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://glasnaya.media/2023/03/13/nakazanie-rodinoj-aeroflot-vnov-okazalsya-v-centre-diskriminacionnogo-skandala/"],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json]: better layout

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
 
 
 def main() -> None:
     """
     Entrypoint for scrapper module
     """
-    pass
-
+    base_path = Path('C:\\pyproj\\hse\\2022-2-level-ctlr\\tmp\\articles')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: use ASSETS_PATH from constants, otherwise it never works on CI side

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            if txt != '\xa0':
+                result = ' '.join([result, txt])
+        self.article.text = result
 
     def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> None:
         """
         Finds meta information of article
         """
-        pass
+        title_bs = article_soup.find('h1', {'class': 'elementor-heading-title elementor-size-default'})
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: is there a better way to find date? sequence of classes is fragile. (here and further)

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        date_bs = article_soup.find('span',
+                                    {'class': 'elementor-icon-list-text elementor-post-info__item elementor-post-info__item--type-date'})
+        date_txt = re.search(r'\d{2}/\d{2}/\d{4}', date_bs.text)
+        self.article.date = date_txt[0]
+
+        auth_bs = article_soup.find('span', {
+            'class': 'elementor-icon-list-text elementor-post-info__item elementor-post-info__item--type-author'})
+        auth_txt = re.search(r'\w+\s\w+', auth_bs.text)
+        self.article.author = auth_txt[0]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: author should be a list

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        topic_bs = article_soup.find('h3', {'class': 'elementor-heading-title elementor-size-default'})
+        topic_txt = topic_bs.find('span').text
+        self.article.topics = topic_txt
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: topics should be a list

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        self.article.article_id = self.article_id
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: should be defined on init level when we create an instance

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         Finds meta information of article
         """
-        pass
+        title_bs = article_soup.find('h1', {'class': 'elementor-heading-title elementor-size-default'})
+        self.article.title = title_bs.text
+
+        date_bs = article_soup.find('span',
+                                    {'class': 'elementor-icon-list-text elementor-post-info__item elementor-post-info__item--type-date'})
+        date_txt = re.search(r'\d{2}/\d{2}/\d{4}', date_bs.text)
+        self.article.date = date_txt[0]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: date should be a datetime instance, and you have to call your unify_date_format method

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        self.article = Article(url=full_url, article_id=article_id)
 
     def _fill_article_with_text(self, article_soup: BeautifulSoup) -> None:
         """
         Finds text of article
         """
-        pass
+        texts_bs = article_soup.find('div', {
+            'class':
+                'elementor-element elementor-element-6732342 elementor-widget elementor-widget-theme-post-content'})
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: too fragile

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        texts = texts_bs.find_all('p')
+        result = ''
+        for el in texts:
+            txt = el.text
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: I hope it all can be made with strip or get_text(strip=True)

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -120,46 +254,96 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            txt = txt.replace('\n\t\t\t\t', '')
+            txt = txt.replace('\t\t\t', '')
+            if txt != '\xa0':
+                result = ' '.join([result, txt])
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: join should happen once and outside the loop

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -90,25 +210,39 @@ def __init__(self, config: Config) -> None:
...
+        return article_bs.get('href')
 
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        for element in self.seed_urls:
+            response = make_request(element, self.config)  # making request for each link
+            if response.status_code != 200:  # i'm not sure whether i need this check
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: remove unrelated comments

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -90,25 +210,39 @@ def __init__(self, config: Config) -> None:
...
+                continue
+
+            main_bs = BeautifulSoup(response.text, 'lxml')
+            paragraphs = main_bs.find_all('h1', {'class': 'elementor-heading-title elementor-size-default'}) # searching paragraphs
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: .

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -90,25 +210,39 @@ def __init__(self, config: Config) -> None:
...
+
+            for each_par in paragraphs:
+                ans = each_par.find_all('a')  # searching for all 'a' tags
+                for elem in ans:
+                    self.urls.append(self._extract_url(article_bs=elem))  # getting the link
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: you do not control number of articles that you collect

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -90,25 +210,39 @@ def __init__(self, config: Config) -> None:
...
+        self.config = config
+        self.seed_urls = config.get_seed_urls()
+        self.urls = []
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        return article_bs.get('href')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: who guarantees that there will be a link attribute? you potentially return None, later ignore it, also breaking interface that requires you to always return a string

File: lab_5_scrapper/scrapper.py [hunk 16]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+
+    response = requests.get(url, config.get_headers())
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: pass timeout and verification flag from config

File: lab_5_scrapper/scrapper.py [hunk 17]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
+    time.sleep(config.get_timeout())
+
+    return response
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:207]: set encoding to response before return

File: lab_5_scrapper/scrapper.py [hunk 18]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+
+    response = requests.get(url, config.get_headers())
+    time.sleep(config.get_timeout())
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: config's timeout is for get request, for sleeps you need to generate some random value from interval

File: lab_5_scrapper/scrapper.py [hunk 19]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
         """
-        pass
+        return self.config_obj.timeout
 
     def get_verify_certificate(self) -> bool:
         """
         Retrieve whether to verify certificate
         """
-        pass
+        return self.config_obj.should_verify_certificate
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: getters should return own attributes and not internal state of DTO instance

File: lab_5_scrapper/scrapper.py [hunk 20]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
+        if not isinstance(self._headers, dict):
+            raise IncorrectHeadersError("Headers are not in a form of dictionary")
+
+        if not isinstance(self._encoding, str):
+            raise IncorrectEncodingError("Encoding must be specified as a string")
+
+        if not isinstance(self._timeout, int):
+            raise IncorrectTimeoutError
+
+        if self._timeout <= 0 or self._timeout > 60:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: all constants should be taken from constants

File: lab_5_scrapper/scrapper.py [hunk 21]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
-        pass
+        if not isinstance(self._seed_urls, list):
+            raise IncorrectSeedURLError
+
+        for element in self._seed_urls:
+            if not isinstance(element, str):
+                raise IncorrectSeedURLError
+
+            if 'https://' not in element:
+            #if not re.match(r'https://', element):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: rm unused code

File: lab_5_scrapper/scrapper.py [hunk 22]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
+                raise IncorrectSeedURLError
+
+        if not isinstance(self._num_articles, int)\
+                or isinstance(self._num_articles, bool) or self._num_articles <= 0:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: why do you check for bool??

File: lab_5_scrapper/scrapper.py [hunk 23]
```diff
@@ -13,70 +73,130 @@ def __init__(self, path_to_config: Path) -> None:
...
+        self._validate_config_content()
+        self.config_obj = self._extract_config_content()
 
     def _extract_config_content(self) -> ConfigDTO:
         """
         Returns config values
         """
-        pass
+
+        return ConfigDTO(self._seed_urls,
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: it breaks the whole idea:

in init:

1. you need to extract config's content with a function
2. run validation using DTO from the previuos step
3. fill attributes of config and continue work

**Review by demid5111 — CHANGES_REQUESTED** [review]
you are in the middle of work, good luck and increase your target mark!

File: seminars/main.py
Comment (daryabalba) [reply seminars/main.py:1]: Done

File: seminars/Text_response.txt
Comment (daryabalba) [reply seminars/Text_response.txt:1]: Done

File: seminars/Text_response.html
Comment (daryabalba) [reply seminars/Text_response.html:1]: Done

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: do you mean rewriting it to `self.article.author = [auth_txt[0]]`?

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 8 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: Thank you!

File: lab_5_scrapper/scrapper.py (hunk 11 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 12 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 15 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 16 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 17 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py:207]: done

File: lab_5_scrapper/scrapper.py (hunk 18 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: do I set a sleep with random value here as well? or i should only pass a timeout to the response in this task?

File: lab_5_scrapper/scrapper.py (hunk 19 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 20 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: Done

File: lab_5_scrapper/scrapper.py (hunk 21 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 22 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: because if you pass a `True` value to `isinstance(value, int)`, it will return True, and i don't need the number of articles to be `True` instead of an integer

File: lab_5_scrapper/scrapper.py (hunk 21 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 23 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py [hunk 24]
```diff
@@ -13,70 +75,146 @@ def __init__(self, path_to_config: Path) -> None:
         """
         Initializes an instance of the Config class
         """
-        pass
+        self.path_to_config = path_to_config
+
+        with open(self.path_to_config, 'r', encoding='utf-8') as json_file:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: this should be in extract

File: lab_5_scrapper/scrapper.py [hunk 25]
```diff
@@ -13,70 +75,146 @@ def __init__(self, path_to_config: Path) -> None:
...
+            self.content = json.load(json_file)
+
+        self._validate_config_content()
+
+        self._seed_urls = self.content['seed_urls']
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: fill attributes based on DTO values

File: lab_5_scrapper/scrapper.py [hunk 26]
```diff
@@ -13,70 +75,146 @@ def __init__(self, path_to_config: Path) -> None:
...
+            raise IncorrectVerifyError("Verify certificate value must either be True or False")
 
     def get_seed_urls(self) -> list[str]:
         """
         Retrieve seed urls
         """
-        pass
+        if isinstance(self._seed_urls, list):
+            return self._seed_urls
+        return ['']
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: redundant check. getter should return part of the internal state. otherwise, you fool your user

File: lab_5_scrapper/scrapper.py [hunk 27]
```diff
@@ -13,70 +75,146 @@ def __init__(self, path_to_config: Path) -> None:
...
 
     def get_num_articles(self) -> int:
         """
         Retrieve total number of articles to scrape
         """
-        pass
+        if isinstance(self._num_articles, int):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: remove all checks from getters. hereinafter

File: lab_5_scrapper/scrapper.py [hunk 28]
```diff
@@ -90,25 +228,51 @@ def __init__(self, config: Config) -> None:
...
+        self.seed_urls = config.get_seed_urls()
+        self.urls = []
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        link = article_bs.get('href')
+        if isinstance(link, str):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:230]: check for none should be enough

File: lab_5_scrapper/scrapper.py [hunk 29]
```diff
@@ -120,45 +284,100 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            date_txt = re.search(r'\d{2}/\d{2}/\d{4}', date_bs.text)
+            if date_txt:
+                self.article.date = self.unify_date_format(date_txt[0])
+
+        auth_bs = article_soup.find('li', {'itemprop': 'author'})
+        auth_txt = re.search(r'\w+\s\w+', auth_bs.text)
+        if auth_txt and isinstance(auth_txt[0], str):
+            self.article.author = [auth_txt[0]]
+        else:
+            self.article.author = ['None']
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: read docs, incorrect default

File: lab_5_scrapper/scrapper.py [hunk 30]
```diff
@@ -120,45 +284,100 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        topic_bs = article_soup.find('h3',
+                                     {
+                                         'class':
+                                             'elementor-heading-title elementor-size-default'
+                                     }
+                                     )
+        topic_txt = topic_bs.find('span').text
+        if isinstance(topic_txt, str):
+            self.article.topics = [topic_txt]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:325]: only one topic per page?

File: lab_5_scrapper/scrapper.py [hunk 31]
```diff
@@ -120,45 +284,100 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
     """
-    pass
+    prepare_environment(ASSETS_PATH)
+    configuration = Config(path_to_config=CRAWLER_CONFIG_PATH)
+    crawler = Crawler(config=configuration)
+    crawler.find_articles()
+    for i, full_url in enumerate(crawler.urls, start=1):
+        parser = HTMLParser(full_url=full_url, article_id=i, config=configuration)
+        article = parser.parse()
+        if isinstance(article, Article):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:456]: why do you check for type?

**Review by demid5111 — CHANGES_REQUESTED** [review]
good work, continue

File: lab_5_scrapper/scrapper.py (hunk 30 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py:325]: Yes

File: lab_5_scrapper/scrapper.py (hunk 27 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: Without them i have problems with my py checks (it says that i'm returning `Any` instead of `int`

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 25 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 29 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 31 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py:456]: the checks do not pass otherwise

File: lab_5_scrapper/scrapper.py [hunk 32]
```diff
@@ -120,46 +271,197 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            'ноября': '11',
+            'декабря': '12',
+        }
+        date_txt = re.search(r'\d{2}/\d{2}/\d{4}', date_bs.text)
+
+        if date_txt:
+            self.article.date = self.unify_date_format(date_txt[0])
+        else:
+            date_txt = re.search(r'\d+\s\w+,\s\d+', date_bs.text)
+            if date_txt and isinstance(date_txt[0], str):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: should be part of unify date method

File: lab_5_scrapper/scrapper.py [hunk 33]
```diff
@@ -13,70 +75,133 @@ def __init__(self, path_to_config: Path) -> None:
...
+        return self._headless_mode
 
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:202]: make sleeps among requests

**Review by demid5111 — CHANGES_REQUESTED** [review]
https://glasnaya.media/2023/03/15/vypusk-48-kak-mobilizaciya-izmenila-rynok-truda-dlya-zhenshhin/ extra text

**Comment by demid5111** [comment]
nothing to review to this moment

File: lab_5_scrapper/scrapper.py (hunk 32 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 33 above)
Comment (daryabalba) [reply lab_5_scrapper/scrapper.py:202]: done

**Comment by vitazaytseva** [comment]
Well done! You can proceed to work on Lab 6)
